src/myrtlespeech/model/transducer.py

```diff
@@ -0,0 +1,584 @@
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+import torch
+
+
+class Transducer(torch.nn.Module):
+    r"""`Transducer <https://arxiv.org/pdf/1211.3711.pdf>`_ Network.
+
+    Args:
+        encoder: A :py:class:`torch.nn.Module` to use as the encoder. Note
+            that this network is also referred to as the Transcription Network
+            in the literature. The ``encoder`` must accept as input a Tuple
+            where the first element is the ``encoder`` input: a
+            :py:class:`torch.Tensor` with size ``[batch, channels,
+            features, max_input_seq_len]`` and the second element is a
+            :py:class:`torch.Tensor` of size ``[batch]`` where each entry
+            represents the sequence length of the corresponding *input*
+            sequence to the rnn.
+
+            It must return a tuple where the first element is the result after
+            applying the module to the input. It must have size
+            ``[max_out_seq_len, batch, output_features]``. The second element
+            of the tuple return value is a :py:class:`torch.Tensor` with size
+            ``[batch]`` where each entry represents the sequence length of the
+            corresponding *output* sequence. These may be different than the
+            input sequence lengths due to downsampling in `encoder`.
+
+            It is possible *but not necessary* to use an initialized
+            :py:class:`TransducerEncoder` class as the `encoder`.
+
+        predict_net: A :py:class:`torch.nn.Module` to use as the prediction
+            network. ``predict_net`` must accept as input a Tuple where the
+            first element is the target label tensor of size ``[batch,
+            max_label_length]`` and the second is a :py:class:`torch.Tensor`
+            of size ``[batch]`` that contains the *input* lengths of these
+            target label sequences.
+
+            It must return a tuple where the first element is the result after
+            applying the module to the input and it must have size ``[batch,
+            max_output_seq_len + 1, in_features]``. Note that the dimension at
+            index 1 is ``max_label_seq_len + 1`` since the start-of-sequence
+            token is prepended to the label sequence. The second element of
+            the Tuple return value is a :py:class:`torch.Tensor` with size
+            ``[batch]`` where each entry represents the sequence length of the
+            corresponding *output* label sequence.
```

Comment:
Good point. I've updated to make clear that the input lengths are returned.

Reply:
There is a question here regarding whether the lengths *should* be returned with +1. 
* Pro: this is probably closer to what the user will expect and allows down sampling in the decoder (although I'm not sure if the latter is useful with language token inputs). 
* Con: the rnnt loss requires the true lengths rather than the +1 lengths. 
I've gone with not adding +1 but am happy to change this if you prefer the other way around?